aten/src/ATen/core/type.cpp

```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+    for (size_t j = std::min(i-1, end_idx); ; --j) {
+      c10::optional<TypePtr> unified = wrapper(types[i], types[j]);
+      if (unified) {
+        types[j] = *unified;
+        --end_idx;
+        break;
+      }
+      // Break condition here so we don't get `j = 0; j = j-1` and end
+      // up with MAX_INT
+      if (j == 0) {
+        break;
+      }
+    }
+  }
+  // Cut off the vector's tail so that `end` is the real last element
+  types.erase(types.begin() + end_idx + 1, types.end());
+}
+
+void standardizeUnion(std::vector<TypePtr>& types) {
+  filterDuplicateSubtypes(types);
+
+  // We want the elements to be sorted so we can easily compare two
+  // UnionType objects for equality in the future. Note that this order
+  // is guaranteed to be stable since we've already coalesced any
+  // possible types
+  std::sort(types.begin(), types.end(),
+          [](const TypePtr a, const TypePtr b) -> bool {
+            if (a->kind() != b->kind()) {
+              return a->kind() < b->kind();
+            }
+            return a->str() < b->str();
+          });
+}
+
+UnionType::UnionType(std::vector<TypePtr> types, TypeKind kind) : Type(kind) {
+  for (auto type : types) {
+    flattenUnion(type, types_);
+  }
+
+  standardizeUnion(types_);
+
+  TORCH_INTERNAL_ASSERT(types_.size() != 1, "After type unification was"
+                        " performed, the Union only has one type. "
+                        "Use the common supertype instead of creating "
+                        "a Union type");
+
+  TORCH_INTERNAL_ASSERT(types_.size() >= 2, "Cannot create a Union of "
+                        "one or fewer types");
+
+  can_hold_none_ = false;
+  has_free_variables_ = false;
+
+  for (const TypePtr type : types_) {
+    if (type->kind() == NoneType::Kind) {
+      can_hold_none_ = true;
+    }
+    if (type->hasFreeVariables()) {
+      has_free_variables_ = true;
+    }
+  }
+
+}
+
+UnionTypePtr UnionType::create(std::vector<TypePtr> types) {
+  auto union_type = new UnionType(std::move(types));
+  if (auto optional_type = union_type->to_optional()) {
+    return OptionalType::create(std::move(*optional_type));
+  }
+  return UnionTypePtr(std::move(union_type));
+}
+
+bool UnionType::operator==(const Type& rhs) const {
+  if (auto union_rhs = rhs.cast<UnionType>()) {
+    return types_ == union_rhs->types_;
+  }
+  if (auto optional_rhs = rhs.cast<OptionalType>()) {
+    auto optional_lhs = this->to_optional();
+    if (optional_lhs) {
+      return *optional_rhs == *((optional_lhs.value())->cast<OptionalType>());
+    }
+    return false;
+  }
+  return false;
+}
+
+bool UnionType::isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const {
+  if (const auto union_rhs = rhs_->cast<UnionType>()) {
+    // Fast path: The Union `types` vector is always sorted, so we can
+    // do an O(n) linear comparison of both vectors. We need to use
+    // `std::equal` with `TypeEqual` for optimal performance; do not
```

Comment:
We're bypassing the `==` handling of Union/Optionals here because we are checking kind